Draft: Get started guide xk6 browser

mdcruz opened the pull request.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/k6-docs?pullRequest=828) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/k6-docs?pullRequest=828) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/k6-docs?pullRequest=828) it.</sub>

inancgumus commented on src/data/markdown/docs/30 xk6-browser/01 Get started/02 Running xk6-browser.md:
```diff
@@ -0,0 +1,236 @@
+    page.$('input[name="password"]').type('123');
+
+    // Wait for asynchronous operations
+    // eslint-disable-next-line no-undef
```
IMO, this comment can confuse users. Should we explain it or write `Promise.all` differently?

mdcruz replied:
`<!-- eslint-skip -->` works now so I've remove the eslint single line comment :)

inancgumus commented on src/data/markdown/docs/30 xk6-browser/01 Get started/02 Running xk6-browser.md:
```diff
@@ -0,0 +1,236 @@
+
+## Run both browser-level and protocol-level in a single script
+
+The real power of xk6-browser shines when it’s combined with the existing features of k6. A common scenario that you can try is to mix browser-level tests with a protocol-level test which can simulate how your website responds to various performance events.
```
🤩

inancgumus reviewed (commented):
Nice work 👏  I have some suggestions.

mdcruz commented on src/data/markdown/docs/30 xk6-browser/01 Get started/02 Running xk6-browser.md:
```diff
@@ -0,0 +1,236 @@
+
+## Interact with elements on your webpage
+
+To interact with elements on the page, you need to use `page.locator()` and pass in the selector of the element. You can also use `page.$()` instead of `page.locator()`.
+
+This will create and return a [Locator](https://k6.io/docs/javascript-api/xk6-browser/locator/) object.
```
This is clearer to me 👍 thanks!

mdcruz commented on src/data/markdown/docs/30 xk6-browser/01 Get started/02 Running xk6-browser.md:
```diff
@@ -0,0 +1,236 @@
+  
+    // Enter login credentials and click submit button
+    page.locator('input[name="login"]').type('admin');
+    page.$('input[name="password"]').type('123');
```
Yeah, I think now that you've suggested the note regarding `page.$`, I'll keep the example as `page.locator` for consistency.

mdcruz commented on src/data/markdown/docs/30 xk6-browser/01 Get started/02 Running xk6-browser.md:
```diff
@@ -0,0 +1,236 @@
+
+  </CodeGroup>
+
+The preceding code creates and returns a Locator object with the selectors for both login and password passed as arguments. 
```
Resolved as of this [commit](https://github.com/grafana/k6-docs/pull/828/commits/67649cb3b9ca4e347bf52a9fead9e354fa5383f0)

inancgumus commented on src/data/markdown/docs/30 xk6-browser/01 Get started/02 Running xk6-browser.md:
```diff
@@ -118,11 +120,11 @@ To avoid timing errors or other race conditions in your script, if you have acti
     // Wait for asynchronous operations
     // eslint-disable-next-line no-undef
-    Promise.all([page.waitForNavigation(), page.$('input[type="submit"]').click()])
+    Promise.all([page.waitForNavigation(), page.locator('input[type="submit"]').click()])
```
The page locator does not support async click yet :-( So it is valid to use `page.$` here with a comment explaining the locator.click does not yet support async click.

mdcruz replied:
Oh, I didn't know that, thanks for clarifying! Will change it back 👍

mdcruz commented on src/data/markdown/docs/30 xk6-browser/01 Get started/02 Running xk6-browser.md:
```diff
@@ -0,0 +1,236 @@
+
+  </CodeGroup>
+
+The preceding code uses `Promise.all([])` to wait for the two promises to be resolved before continuing. Then, you can use [`check`](https://k6.io/docs/javascript-api/k6/check/) from the k6 API to assert the text content of a specific element. Finally, you close the page and the browser.
```
Will add 👍

mdcruz commented on src/data/markdown/docs/30 xk6-browser/01 Get started/02 Running xk6-browser.md:
```diff
@@ -0,0 +1,236 @@
+    const browser = chromium.launch({ headless: false });
+    const page = browser.newPage();
+
+    page.goto('https://test.k6.io/my_messages.php', { waitUntil: 'networkidle' });
```
There's also other examples in the doc (not just in this PR) that we need to change so will change those too.

mdcruz replied:
I have now made the goto API async for the examples in this PR. There are also other examples outside the get started guide (out of scope for this PR) that we need to create as a separate PR.

ppcano commented:
@bolotskydev, related to the changes in this PR.


<img width="525" alt="Screenshot 2022-10-17 at 12 18 22" src="https://user-images.githubusercontent.com/825430/196153096-111085d5-a353-4e38-be8b-876ea2471322.png">


In the [`xk6-browser` documentation](https://mdr-ci.staging.k6.io/docs/refs/pull/828/merge/javascript-api/xk6-browser/), I'd like the `Welcome/Overview` page to be shown on the Sidebar.  As we do in the `Guides` section  👇👇


<img width="521" alt="Screenshot 2022-10-17 at 12 18 03" src="https://user-images.githubusercontent.com/825430/196153070-30dcdd58-d717-486e-8c4b-117bcc22aae5.png">

cc @mdcruz

ppcano commented on src/data/markdown/docs/30 xk6-browser/01 Get started/02 Running xk6-browser.md:
```diff
@@ -0,0 +1,245 @@
+
+You can add assertions in your browser script via:
+1. Checks, as mentioned in the previous section
+2. [k6chaijs](https://k6.io/docs/javascript-api/jslib/k6chaijs/) for a more BDD (Behavior Driven Development) or TDD (Test Driven Development) style
```
@mdcruz I think we should not recommend `k6chaijs` at this moment and stick with "official" APIs. 

We have very little documentation about k6chaijs. The 99% of our assertion examples use Checks. 

Introducing `k6chaijs` in this article adds additional overhead.

mdcruz replied:
Makes sense Pepe, I'll remove this section then 👍

ppcano commented on src/data/markdown/docs/30 xk6-browser/01 Get started/03 Metrics.md:
```diff
@@ -0,0 +1,14 @@
+---
+title: 'Browser performance metrics'
```
Alternative titles:
- Browser metrics
- Built-in metrics
- Browser built-in metrics
- Metrics

ppcano commented on src/data/markdown/docs/30 xk6-browser/01 Get started/03 Metrics.md:
```diff
@@ -0,0 +1,14 @@
+
+Apart from the usual HTTP specific metrics that k6 already tracks, xk6-browser tracks the following browser specific performance metrics on top:
+
+- browser_dom_content_loaded - measures the time for when the page is loaded but scripts have just started to be executed.
```
Can we use a table here to be consistent on how we represent other metric lists?

See example:
https://k6.io/docs/using-k6/metrics/#http-specific-built-in-metrics

ppcano commented on src/data/markdown/docs/30 xk6-browser/01 Get started/03 Metrics.md:
```diff
@@ -0,0 +1,14 @@
+- browser_first_meaningful_paint - measures the time for when the primary content, which is meaningful, is painted on the browser.
+- browser_loaded - measures the time for when the page is fully loaded.
+
+> Note that the above metrics are still subject to change in the future.
```
Warning?

mdcruz replied:
Will change 👍

ppcano commented on src/data/markdown/docs/30 xk6-browser/01 Get started/04 Module Properties.md:
```diff
@@ -0,0 +1,38 @@
+---
+title: 'Module Properties'
```
This seems docs about the API. Don't know how but I'd place it within the Module section.

<img width="274" alt="Screenshot 2022-10-17 at 13 12 51" src="https://user-images.githubusercontent.com/825430/196163133-18ec9359-7f8e-4427-9974-8764cd80be54.png">

mdcruz replied:
I've created 
<img width="271" alt="Screenshot 2022-10-17 at 17 56 14" src="https://user-images.githubusercontent.com/10811379/196237818-98c3da58-5f4d-47ae-8c14-96f0a4b2873b.png">

ppcano commented on src/data/markdown/docs/30 xk6-browser/30 xk6-browser.md:
```diff
@@ -1,144 +1,23 @@
+- When my application is receiving thousands of simultaneous requests from the protocol-level, what happens to the front-end?
+- How can I get metrics specific to browsers, like total page load time?
+- Are all my elements interactive on the front-end?
+- Are there any loading spinners that take a long time to disappear?
```
General feedback on the Intro/Welcome page.

- I think it should mention that xk6-browser mimics the Playwright API.
- Show a simple example to help users get a mental model of what looks like.

mdcruz replied:
I've added this [commit](https://github.com/grafana/k6-docs/pull/828/commits/a7c3614e642a1c17194e4353669b57e399f35b1a) Pepe

inancgumus commented on src/data/markdown/docs/30 xk6-browser/01 Get started/02 Running xk6-browser.md:
```diff
@@ -0,0 +1,236 @@
+
+Within the Locator API, various methods such as `type()` or `click()` can be used to interact with the elements. The `type()` method types a text to an input field while `click()` will do a mouse click on the element. 
+
+## Asynchronous operations
```
The thing is `page.goto` will be async in the next release (it's in the current `main`) :(

mdcruz replied:
Shall we wait and get this merge in the next release as well? 🤔

inancgumus replied:
It seems like you updated the examples. So I'm resolving this suggestion now 👍

mdcruz commented:
> <img alt="Screenshot 2022-10-17 at 13 14 41" width="927" src="https://user-images.githubusercontent.com/825430/196163446-766a457c-490d-4a1e-88fe-e144fc468eb5.png">
> 
> @mdcruz imo, the content of Browser Level APIs should be for the xk6-browser page.
> 
> <img alt="Screenshot 2022-10-17 at 13 16 07" width="583" src="https://user-images.githubusercontent.com/825430/196163843-0f3b961c-7cdb-4f72-a5c6-4aa58b13ef4f.png">
> 
> For this, we'll need a Welcome/Overview page that I requested [#828 (comment)](https://github.com/grafana/k6-docs/pull/828#issuecomment-1280635750). wdyt?

Yup, we would need the welcome/overview page for this. The breadcrumbs for xk6-browser is also broken in this PR which I think should be fixed by [#828 (comment)](https://github.com/grafana/k6-docs/pull/828#issuecomment-1280635750)

bolotskydev commented:
@ppcano @mdcruz in latest commit I've added a `Welcome` page for xk6-browser as requested. Please, confirm it is indeed what you've expected.

⚠️  A couple of notes: 
- Breadcrumbs don't work and shouldn't - they were disabled altogether for `javascript-api` during versioning introduction. Since we revamped the section, I'm working on enabling them for `xk6-browser`.
- Please acknowledge that I've renamed `Get started` -> `Getting started`, for the sake of consistency (see `Guides`). In case it is unwanted change, let me know, I will roll it back

mdcruz commented:
> @ppcano @mdcruz in latest commit I've added a `Welcome` page for xk6-browser as requested. Please, confirm it is indeed what you've expected.
> 
> ⚠️ A couple of notes:
> 
> * Breadcrumbs don't work and shouldn't - they were disabled altogether for `javascript-api` during versioning introduction. Since we revamped the section, I'm working on enabling them for `xk6-browser`.
> * Please acknowledge that I've renamed `Get started` -> `Getting started`, for the sake of consistency (see `Guides`). In case it is unwanted change, let me know, I will roll it back

Thanks @bolotskydev for the speedy change! ❤️ 

So, for now, it's expected for the breadcrumbs to link to a 404 page? https://mdr-ci.staging.k6.io/docs/refs/pull/828/merge/javascript-api/xk6-browser/api-overview/

Yup, let's change it to Getting started for consistency. Thanks!

bolotskydev commented:
@mdcruz yep, correct, user will end up on 404, as currently no stub breadcrumb pages are generated for `javascript-api` subroutes. I think we either let it be while I am working on patch, or disable breadcrumbs for this section altogether

ppcano commented:
@bolotskydev  I applied a few changes. But we still need additional changes:


1 - The `xk6-browser` welcome page should be the root of the subdirectory. Available under `/javascript-api/xk6-browser`


- The header menu option `JAVASCRIPT API / XK6-BROWSER` should point to the location of the Welcome page (`/javascript-api/xk6-browser`).


2 - All xk6-browser APIs should be available under `xk6-browser/apis`. 
The sidebar menu should look the same. We can create the necessary redirections, later.



3 - `/javascript-api/xk6-browser/getting-started/` is 404

4 - The dropdown sidebar menu does not select the `xk6-browser` option for all the pages.

5 - On the `Welcome` page. We should remove the h1 element -> `Welcome`. To be consistent with other pages, we could also remove the breadcrumb.

imiric commented on src/data/markdown/docs/30 xk6-browser/01 Getting started/03 Running xk6-browser.md:
```diff
@@ -0,0 +1,227 @@
+
+## Asynchronous operations
+
+As explained previously, k6 operations are synchronous. However, Playwright operations are asynchronous in nature and since xk6-browser aims to provide rough compatibility with Playwright, some of our operations are now asynchronous.
```
It might be good to mention that we're not just transitioning to async to follow Playwright's lead, but because it's a better fit for interacting with the browser.

Also, let's stick to a single way of referring to functions, or methods in this case, instead of "operations", which is a bit ambiguous in this context.

```suggestion
As explained previously, the k6 API is synchronous. However, since many browser operations happen asynchronously, and in order to follow the Playwright API more closely, we are working on migrating most xk6-browser methods to be asynchronous as well.
```

imiric commented on src/data/markdown/docs/30 xk6-browser/01 Getting started/03 Running xk6-browser.md:
```diff
@@ -0,0 +1,227 @@
+
+As explained previously, k6 operations are synchronous. However, Playwright operations are asynchronous in nature and since xk6-browser aims to provide rough compatibility with Playwright, some of our operations are now asynchronous.
+
+At the moment, operations such as `page.goto()`, `page.waitForNavigation()` and `Element.click()` return promises and work similarly to [JavaScript promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises).
```
```suggestion
At the moment, methods such as `page.goto()`, `page.waitForNavigation()` and `Element.click()` return promises and work similarly to [JavaScript promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises).
```

imiric commented on src/data/markdown/docs/30 xk6-browser/01 Getting started/04 Metrics.md:
```diff
@@ -0,0 +1,21 @@
+
+| Metric Name                    |  Description                                                                                     |
+|--------------------------------|--------------------------------------------------------------------------------------------------|
+| browser_dom_content_loaded     | Measures the time for when the page is loaded but scripts have just started to be executed.      |
+| browser_first_paint            | Measures the time for when the first pixel is painted on the browser.                            |
+| browser_first_contentful_paint | Measures the time for when the first bit of content is painted on the browser.                   |
+| browser_first_meaningful_paint | Measures the time for when the primary content, which is meaningful, is painted on the browser.  |
+| browser_loaded                 | Measures the time for when the page is fully loaded.                                             |
```
"Measures the time for when" sounds a bit awkward, and the metric itself doesn't "measure" anything.

So maybe:

```suggestion
| browser_dom_content_loaded     | Emitted when the page is loaded but scripts have just started to be executed.      |
| browser_first_paint            | Emitted when the first pixel is painted on the browser.                            |
| browser_first_contentful_paint | Emitted when the first bit of content is painted on the browser.                   |
| browser_first_meaningful_paint | Emitted when the primary content, which is meaningful, is painted on the browser.  |
| browser_loaded                 | Emitted when the page is fully loaded.                                             |
```

imiric commented on src/data/markdown/docs/30 xk6-browser/01 Getting started/04 Metrics.md:
```diff
@@ -0,0 +1,21 @@
+| Metric Name                    |  Description                                                                                     |
+|--------------------------------|--------------------------------------------------------------------------------------------------|
+| browser_dom_content_loaded     | Measures the time for when the page is loaded but scripts have just started to be executed.      |
+| browser_first_paint            | Measures the time for when the first pixel is painted on the browser.                            |
+| browser_first_contentful_paint | Measures the time for when the first bit of content is painted on the browser.                   |
```
Hhmm the explanation of these two sounds very similar. I'm honestly not sure about the difference myself, so can we look up a better description that makes it clearer what the difference is?

mdcruz replied:
Will change these, thanks Ivan!

imiric commented on src/data/markdown/docs/30 xk6-browser/01 Getting started/04 Metrics.md:
```diff
@@ -0,0 +1,21 @@
+| browser_dom_content_loaded     | Measures the time for when the page is loaded but scripts have just started to be executed.      |
+| browser_first_paint            | Measures the time for when the first pixel is painted on the browser.                            |
+| browser_first_contentful_paint | Measures the time for when the first bit of content is painted on the browser.                   |
+| browser_first_meaningful_paint | Measures the time for when the primary content, which is meaningful, is painted on the browser.  |
```
Here as well. "which is meaningful" doesn't explain much, so maybe can we expand on what we mean by "primary content"?

mdcruz replied:
Will change this, thanks Ivan!

imiric reviewed (changes requested):
Great work, Marie! :clap:

I'll give this another look later, but here are some initial suggestions.

mdcruz commented on src/data/markdown/docs/30 xk6-browser/01 Getting started/01 Welcome.md:
```diff
@@ -0,0 +1,121 @@
+
+While backend performance testing is useful and efficient, it doesn’t represent the overall user experience. 
+
+With xk6-browser, you can interact with the browser to script tests much closer to the actual website experience and load test the backend to get the complete picture. You'll get insight into what your user sees on a browser-level and catch issues skipped from the protocol level.
```
Ah, I think I didn't make it clearer! You explained it much better than me :D 

My intention was to say that xk6-browser can do the browser performance check while k6 can do the backend performance test.

I think it's worth though to have a best practice guide (something that I'll work with @nicolevanderhoeven to say that people shouldn't use xk6-browsers to launch thousands of VUs concurrently.

ppcano commented:
@bolotskydev 


2 - 

> What is the supposed context for this page? the one we currently have [here](https://mdr-ci.staging.k6.io/docs/refs/pull/828/merge/javascript-api/xk6-browser/)?

Yes

> Should we redirect the page I've just mentioned to /apis ?

Yes but to `/api`. Corrected below.

> Are your sure the apis should be in plural?

I double-checked it. Let's use the singular -> `xk6-browser/api`

-------

3 - I created the markdown file and it works. Should be fixed by https://github.com/grafana/k6-docs/pull/828/commits/a01a378c65036d9548065bf1b56e2741377dbb42

ankur22 commented on src/data/markdown/docs/30 xk6-browser/01 Getting started/01 Welcome.md:
```diff
@@ -0,0 +1,123 @@
+
+The [performance golden rule](https://www.stevesouders.com/blog/2012/02/10/the-performance-golden-rule/) states that 80%-90% of the website response time is spent on the frontend. But when the number of concurrent users is high, the percentage of time spent between the frontend and backend changes:
+
+![Frontend time vs Backend time with increasing amount of website visitors](./images/Frontend-Backend-LoadTesting.png)
```
The image isn't displaying for me in the welcome page. Not sure why though :/
<img width="403" alt="Screenshot 2022-10-20 at 16 02 48" src="https://user-images.githubusercontent.com/1112428/196986332-8b5bdafa-334e-480f-bec8-bcffb77e523e.png">

mdcruz replied:
Yeah, in localhost it doesn't get displayed :( But if I deploy to Staging, it's displayed 🤷🏻‍♀️

https://mdr-ci.staging.k6.io/docs/refs/pull/828/merge/javascript-api/xk6-browser/getting-started/welcome/

This is also the same image used in https://mdr-ci.staging.k6.io/docs/refs/pull/828/merge/testing-guides/load-testing-websites/ and it also doesn't display locally.

ankur22 commented on src/data/markdown/docs/30 xk6-browser/01 Getting started/02 Installation.md:
```diff
@@ -0,0 +1,36 @@
+excerpt: 'A step-by-step guide on how to install xk6-browser.'
+---
+
+xk6-browser is currently being developed as a [k6 extension](/extensions). You have to run a k6 version built with the browser extension to use the [browser-level APIs](#browser-level-apis) in your k6 tests.
```
This link (`browser-level APIs`) doesn't work. I think it would be good to note here that the instructions on how to build a k6 version with the browser extension are below (or just link to it). People who are curious will click on it, and people who don't care about the details can ignore it.

mdcruz replied:
I'll just move the instructions below after this paragraph :)

ankur22 commented on src/data/markdown/docs/30 xk6-browser/01 Getting started/02 Installation.md:
```diff
@@ -0,0 +1,36 @@
+go install go.k6.io/xk6/cmd/xk6@latest
+
+# Build the xk6-browser binary
+xk6 build --output xk6-browser --with github.com/grafana/xk6-browser
```
On Windows this currently needs to be `xk6 build --output xk6-browser.exe --with github.com/grafana/xk6-browser` (NOTE: the `.exe` extension).

Would be good if we had Windows specific instruction here, like we do [here](https://k6.io/docs/getting-started/running-k6/#adding-more-vus).

ankur22 replied:
We have an open ticket for this, so maybe it's worth doing it separately? https://github.com/grafana/xk6-browser/issues/592

ankur22 replied:
Also, would it be worth suggesting that the user uses `xk6 build --output xk6-browser --with github.com/grafana/xk6-browser@latest` which will be the up to date version on the `main` branch, vs `xk6 build --output xk6-browser --with github.com/grafana/xk6-browser` which will build the last tagged version?

mdcruz replied:
I don't mind adding the same instructions on Windows in this PR as there are still outstanding changes required from our external design team.

Regarding `xk6 build --output xk6-browser --with github.com/grafana/xk6-browser@latest`, I asked about this before during one of our team catch-ups and I think the reply was we want our users to use the latest release version because it's the stable version. @imiric WDYT?

imiric replied:
@ankur22 I don't think that's right.

If `@latest` is not specified, xk6 will append it, so whether the user specifies it or not doesn't matter. And `@latest` will build the latest released/tagged version, not the latest `main` commit. This is why we often suggest internal testers and users who want the latest _unstable_ version to install with `@main` instead. So leaving this as is is fine. I don't think we even need to document the `@main` command, just to avoid confusing users more.

As for appending `.exe` on Windows, I agree and think we should mention it here, and anywhere else we document `xk6 build`.

ankur22 replied:
@imiric thanks for clarifying 👍

mdcruz replied:
Addressed as part of https://github.com/grafana/k6-docs/pull/828/commits/9aafb34800a1dad3e764367391e28451ad0b6bd2

ankur22 commented on src/data/markdown/docs/30 xk6-browser/01 Getting started/03 Running xk6-browser.md:
```diff
@@ -0,0 +1,235 @@
+    const page = browser.newPage();
+
+    page
+      .goto('https://test.k6.io/')
```
```suggestion
      .goto('https://test.k6.io/', { waitUntil: 'networkidle' })
```

mdcruz commented on src/data/markdown/docs/30 xk6-browser/01 Getting started/03 Running xk6-browser.md:
```diff
@@ -0,0 +1,235 @@
+
+You can use `page.locator()` and pass in the element's selector you want to find on the page. `page.locator()` will create and return a [Locator](/javascript-api/xk6-browser/locator/) object, which you can later use to interact with the element.
+
+To find out which selectors xk6-browser supports, check out [Selecting Elements](/javascript-api/xk6-browser/get-started/selecting-elements/).
```
Ah yes, will fix!

mdcruz commented on src/data/markdown/docs/30 xk6-browser/01 Getting started/02 Installation.md:
```diff
@@ -0,0 +1,36 @@
+
+xk6 will create the `xk6-browser` binary in the current working directory.
+
+To find out more about how to use xk6 or what it is, check out this guide - [Build a k6 binary with extensions](https://k6.io/docs/extensions/guides/build-a-k6-binary-with-extensions/).
```
@ppcano @bolotskydev Another one for the design changes?

bolotskydev commented:
@ppcano @mdcruz I'll be checking for regressions on staging, but feel free to explore the latest changes already

ppcano commented:
@bolotskydev 


![Screenshot 2022-11-04 at 10 35 45](https://user-images.githubusercontent.com/825430/199949806-2db355e1-008c-4506-8221-0214b65850fc.png)

1 - The breadcrumb link points to the root instead to `xk6-browser/api`.


2 - Also, I think it is confusing not to have the xk6-browser APIs under `xk6-browser/api`

Instead of `xk6-browser/browser` --> `xk6-browser/api/browser`. 

This breaks the existing pattern, but I see it as more intuitive.  It seems we have two distinct API project types:

- javascript API and jslib --> Only APIs
- xk6-browser, xk6-disruptor -> Guides + APIs

ankur22 reviewed (approved):
Nice work @mdcruz! LGTM

inancgumus commented on src/data/markdown/docs/30 xk6-browser/01 Get started/03 Running xk6-browser.md:
```diff
@@ -0,0 +1,237 @@
+        check(page, {
+          'header': page.locator('h2').textContent() == 'Welcome, admin!',
+        });
+      }).finally(() => {
+        page.close();
+        browser.close();
+      });
+    });
```
The `finally` better be after the first `then` (`.goto`'s `then`) since `goto` can also throw an exception.

```suggestion
      });
    }).finally(() => {
      page.close();
      browser.close();
    });
```

ankur22 replied:
Aha, i was also wondering what was best practice here, thanks for this @inancgumus!

inancgumus commented on src/data/markdown/docs/30 xk6-browser/02 xk6-browser.md:
```diff
@@ -0,0 +1,124 @@
+      check(page, {
+        'header': page.locator('h2').textContent() == 'Welcome, admin!',
+      });
+    }).finally(() => {
+      page.close();
+      browser.close();
+    });
+  });
```
The `finally` better be after the first `then` (`.goto`'s `then`) since `goto` can also throw an exception.

```suggestion
    });
  }).finally(() => {
    page.close();
    browser.close();
  });
```

inancgumus reviewed (approved):
Awesome work! Thank you for your hard work and clear explanations 🙇  This will allow users to get started using k6-browser in no time!

ankur22 commented on src/data/markdown/docs/30 xk6-browser/03 API.md:
```diff
@@ -97,36 +20,32 @@ Listed in the table are the importable properties from the top level module (`'k
-  page.goto('https://test.k6.io/browser.php/', {
-    waitUntil: 'networkidle',
-  });
+    page.goto('https://test.k6.io/', {
```
```go
    page.goto('https://test.k6.io/', { waitUntil: 'networkidle' })
    .finally(() => {
      page.close();
      browser.close();
    });
```

imiric commented on src/data/markdown/docs/30 xk6-browser/01 Getting started/01 Welcome.md:
```diff
@@ -0,0 +1,121 @@
+
+<Blockquote mod="note" title="">
+
+Even though xk6-browser aims to provide rough compatibility with Playwright, xk6-browser currently doesn't support async/await keywords. For now, you would have to use Promises. For more information, check out [Running xk6-browser](/javascript-api/xk6-browser/get-started/running-xk6-browser/).
```
@mdcruz This is still unchanged. Consider rephrasing it to something like:

> We're currently (Nov. 2022) migrating most xk6-browser APIs to be async and return a Promise. Because k6 doesn't yet support the `async`/`await` keywords, you'll have to use `then()` to resolve Promises. For more information, check out [Running xk6-browser](/javascript-api/xk6-browser/get-started/running-xk6-browser/).

mdcruz replied:
Thanks Ivan! I'll update it :)

imiric commented on src/data/markdown/docs/30 xk6-browser/01 Get started/03 Running xk6-browser.md:
```diff
@@ -0,0 +1,237 @@
+---
+title: 'Running xk6-browser'
+excerpt: 'Follow along to learn how to run xk6-browser test, interact with elements on the page, wait for page navigation, write assertions and run both browser-level and protocol-level in a single script.'
```
```suggestion
excerpt: 'Follow along to learn how to run xk6-browser test, interact with elements on the page, wait for page navigation, write assertions and run both browser-level and protocol-level tests in a single script.'
```

imiric commented on src/data/markdown/docs/30 xk6-browser/01 Get started/03 Running xk6-browser.md:
```diff
@@ -0,0 +1,237 @@
+1. Run a test
+2. Interact with elements on your webpage
+3. Wait for page navigation
+4. Run both browser-level and protocol-level test in a single script
```
```suggestion
4. Run both browser-level and protocol-level tests in a single script
```

imiric commented on src/data/markdown/docs/30 xk6-browser/01 Get started/03 Running xk6-browser.md:
```diff
@@ -0,0 +1,237 @@
+
+Then, you can use [`check`](/javascript-api/k6/check/) from the k6 API to assert the text content of a specific element. Finally, you close the page and the browser.
+
+## Run both browser-level and protocol-level in a single script
```
```suggestion
## Run both browser-level and protocol-level tests in a single script
```

imiric reviewed (approved):
Just a few minor leftover things, and [a suggestion to change the paragraph about async](https://github.com/grafana/k6-docs/pull/828#discussion_r1015270406).

LGTM otherwise. Great work! :clap:

ankur22 commented on src/data/markdown/docs/30 xk6-browser/02 xk6-browser/10-page.md:
```diff
@@ -71,33 +71,41 @@ excerpt: "xk6-browser: Page Class"
-
-  page.close();
-  browser.close();
+  page.goto('https://test.k6.io/', { waitUntil: 'networkidle' }).then(() => {
```
Minor: Maybe place the `.then(() => {` on a new line like the other examples?
